falcon/hooks.py

```diff
@@ -20,29 +20,94 @@
 from inspect import getmembers
 from inspect import iscoroutinefunction
 import re
-import typing as t
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    cast,
+    Dict,
+    List,
+    Tuple,
+    TYPE_CHECKING,
+    TypeVar,
+    Union,
+)
 
+from falcon._typing_extensions import Protocol
 from falcon.constants import COMBINED_METHODS
 from falcon.util.misc import get_argnames
 from falcon.util.sync import _wrap_non_coroutine_unsafe
 
-if t.TYPE_CHECKING:  # pragma: no cover
+if TYPE_CHECKING:  # pragma: no cover
     import falcon as wsgi
     from falcon import asgi
+    from falcon.typing import AsyncResponderMethod
+    from falcon.typing import Resource
+    from falcon.typing import Responder
+    from falcon.typing import SyncResponderMethod
+
+
+class SyncBeforeFn(Protocol):
+    def __call__(
+        self,
+        req: wsgi.Request,
+        resp: wsgi.Response,
+        resource: Resource,
+        params: Dict[str, Any],
+        *args: Any,
```

Comment:
since we can't use paramspec due to the is_async kw, we can't use a simple `Callable[[request, etc], None]` here since it does not support args and kwargs

Reply:
left a note